## Networks Centrality Measures
In this section, we want to answer our research quenstion by measuring the different centrality measures and understanding the meaning of the results.

### Imports

In [1]:
import networkx as nx
import matplotlib.pylab as plt
import numpy as np
import pandas as pd
import powerlaw
from statistics import mean

### Defining graphs

In [2]:
# Importing the regular graph, for research question no. 1
basic_edges_df = pd.read_csv("basic_graph_edges.csv")
G = nx.from_pandas_edgelist(basic_edges_df,"Source","Target",create_using=nx.DiGraph)
# Importing the graphs of the events, for research question no. 2
G_events={"A":[],"B":[],"C":[]}
before2018_edges_df = pd.read_csv("20180401-20180501_edges_full.csv")
G_events["A"].append(nx.from_pandas_edgelist(before2018_edges_df,"Source","Target",create_using=nx.DiGraph))
after2018_edges_df = pd.read_csv("20180501-20180601_edges_full.csv")
G_events["A"].append(nx.from_pandas_edgelist(after2018_edges_df,"Source","Target",create_using=nx.DiGraph))
before2019_edges_df = pd.read_csv("20190901-20191001_edges_full.csv")
G_events["B"].append(nx.from_pandas_edgelist(before2019_edges_df,"Source","Target",create_using=nx.DiGraph))
after2019_edges_df = pd.read_csv("20191001-20191101_edges_full.csv")
G_events["B"].append(nx.from_pandas_edgelist(after2019_edges_df,"Source","Target",create_using=nx.DiGraph))
before2022_edges_df = pd.read_csv("20220101-20220201_edges_full.csv")
G_events["C"].append(nx.from_pandas_edgelist(before2022_edges_df,"Source","Target",create_using=nx.DiGraph))
after2022_edges_df = pd.read_csv("20220201-20220301_edges_full.csv")
G_events["C"].append(nx.from_pandas_edgelist(after2022_edges_df,"Source","Target",create_using=nx.DiGraph))

### The functions to calculate the measures:

#### Getting Top 5 Maximum Values

In [3]:
def top_five(A):
    return sorted(A, key=A.get, reverse=True)[:5]

#### Degree Centrality

In [4]:
def degree_centrality(G):
    return top_five(nx.degree_centrality(G))

#### In Betweenness

In [5]:
def betweenness(G):
    return top_five(nx.betweenness_centrality(G))

#### In-Degree Centrality

In [6]:
def in_degree (G):
    return top_five(nx.in_degree_centrality(G))

#### Out-Degree Centrality

In [7]:
def out_degree(G):
    return top_five(nx.out_degree_centrality(G))

#### Closeness

In [8]:
def closeness(G):
    return top_five(nx.closeness_centrality(G))

### The basic network

#### Degree Centrelity

In [9]:
degree_centrality(G)

['Ukraine', 'Russian language', 'Poland', 'Ukrainian language', 'World War I']

#### In Betweeness

In [10]:
betweenness(G)

['Ukraine', 'Poland', 'Vladimir Putin', 'Russian language', 'Russians']

#### In-Degree Centrality

In [11]:
in_degree(G)

['Russian language',
 'Ukraine',
 'Ukrainian language',
 'World War I',
 'United States']

#### Out-Degree Centrality

In [12]:
out_degree(G)

['Russian federation', 'Poland', 'Ukraine', 'Ukrainia', 'Soviet']

#### Closeness

In [13]:
closeness(G)

['Russian language', 'Ukraine', 'United States', 'World War I', 'Poland']

### The Special Events Network

#### Degree Centrality

In [14]:
for i in G_events:
    str_to_print = "The nodes with the most centrality before event " + i + " are: "
    print(str_to_print)
    print(degree_centrality(G_events[i][0]))
    str_to_print = "The nodes with the most centrality after event " + i + "  are: "
    print(str_to_print)
    print(degree_centrality(G_events[i][1]))
    print()


The nodes with the most centrality before event A are: 
[1, 515, 904, 1183, 4]
The nodes with the most centrality after event A  are: 
[1, 515, 904, 1183, 4]

The nodes with the most centrality before event B are: 
[1, 515, 904, 1183, 4]
The nodes with the most centrality after event B  are: 
[1, 515, 904, 1183, 4]

The nodes with the most centrality before event C are: 
[1, 515, 904, 4, 1183]
The nodes with the most centrality after event C  are: 
[1, 515, 904, 4, 1183]



Note: 1 represents Ukraine, 515 - Russian language, 904 - Poland, 4 - Ukrainian language, 1183 - World War I

#### In Betweeness

In [15]:
for i in G_events:
    str_to_print = "The nodes with the most betweenness centrality before event " + i + " are: "
    print(str_to_print + str(betweenness(G_events[i][0])))
    str_to_print = "The nodes with the most betweenness centrality after event " + i + " are: "
    print(str_to_print + str(betweenness(G_events[i][1])))
    print()

The nodes with the most betweenness centrality before event A are: [1, 904, 515, 109, 1324]
The nodes with the most betweenness centrality after event A are: [1, 904, 515, 109, 1324]

The nodes with the most betweenness centrality before event B are: [1, 904, 515, 109, 1324]
The nodes with the most betweenness centrality after event B are: [1, 904, 515, 109, 1324]

The nodes with the most betweenness centrality before event C are: [1, 904, 109, 515, 1324]
The nodes with the most betweenness centrality after event C are: [1, 904, 109, 515, 1324]



Note: Every event exept the war in 2022: Ukraine, Poland, Russian Language, Vladimir Putin, Russians.
The war: Ukraine, Poland, Vladimir Putin, Russian Language, Russians.

#### In-Degree Centrality

In [16]:
for i in G_events:
    str_to_print = "The nodes with the most in-degree centrality before event " + i + " are: "
    print(str_to_print + str(in_degree(G_events[i][0])))
    str_to_print = "The nodes with the most in-degree centrality after event " + i + " are: "
    print(str_to_print + str(in_degree(G_events[i][1])))
    print()

The nodes with the most in-degree centrality before event A are: [515, 1, 4, 1183, 4363]
The nodes with the most in-degree centrality after event A are: [515, 1, 4, 1183, 4363]

The nodes with the most in-degree centrality before event B are: [515, 1, 4, 1183, 4363]
The nodes with the most in-degree centrality after event B are: [515, 1, 4, 1183, 4363]

The nodes with the most in-degree centrality before event C are: [515, 1, 4, 1183, 4363]
The nodes with the most in-degree centrality after event C are: [515, 1, 4, 1183, 4363]



Note: Russian language, Ukraine, Ukrainian language, World War I, United States.

#### Out-Degree Centrality

In [17]:
for i in G_events:
    str_to_print = "The nodes with the most out-degree centrality before event " + i + " are: "
    print(str_to_print + str(out_degree(G_events[i][0])))
    str_to_print = "The nodes with the most out-degree centrality after event " + i + " are: "
    print(str_to_print + str(out_degree(G_events[i][1])))
    print()

The nodes with the most out-degree centrality before event A are: [563, 904, 1, 0, 282]
The nodes with the most out-degree centrality after event A are: [563, 904, 1, 0, 282]

The nodes with the most out-degree centrality before event B are: [563, 904, 1, 0, 282]
The nodes with the most out-degree centrality after event B are: [563, 904, 1, 0, 282]

The nodes with the most out-degree centrality before event C are: [563, 904, 1, 0, 282]
The nodes with the most out-degree centrality after event C are: [563, 904, 1, 0, 282]



Note: Russian Federaion, Poland, 1 and 0 are Ukraine, Soviet.

#### Closeness

In [18]:
for i in G_events:
    str_to_print = "The nodes with the most closeness centrality before event " + i + " are: "
    print(str_to_print + str(closeness(G_events[i][0])))
    str_to_print = "The nodes with the most closeness centrality after event " + i + " are: "
    print(str_to_print + str(closeness(G_events[i][1])))
    print()

The nodes with the most closeness centrality before event A are: [515, 1, 4363, 1183, 904]
The nodes with the most closeness centrality after event A are: [515, 1, 4363, 1183, 904]

The nodes with the most closeness centrality before event B are: [515, 1, 4363, 1183, 904]
The nodes with the most closeness centrality after event B are: [515, 1, 4363, 1183, 904]

The nodes with the most closeness centrality before event C are: [515, 1, 4363, 1183, 904]
The nodes with the most closeness centrality after event C are: [515, 1, 4363, 1183, 904]



Note: Russian Language, Ukraine, United States, World War I, Poland.